ethash.go

```diff
@@ -357,82 +258,51 @@ func (pow *Ethash) Search(block pow.Block, stop <-chan struct{}) (uint64, []byte
-			blockNum, epochLength*2048))
-		return false
-	}
-
-	// First check: make sure header, mixDigest, nonce are correct without hitting the cache
-	// This is to prevent DOS attacks
-	chash := (*C.ethash_h256_t)(unsafe.Pointer(&hash[0]))
-	cnonce := C.uint64_t(nonce)
-	target := new(big.Int).Div(minDifficulty, difficulty)
-
-	var pAc *ParamsAndCache
-	// If its an old block (doesn't use the current cache)
-	// get the cache for it but don't update (so we don't need the mutex)
-	// Otherwise, it's the current block or a future block.
-	// If current, updateCache will do nothing.
-	if blockNum/epochLength < pow.paramsAndCache.Epoch {
-		var err error
-		// If we can't make the params for some reason, this block is invalid
-		pAc, err = makeParamsAndCache(pow.chainManager, blockNum)
-		if err != nil {
-			glog.V(logger.Info).Infoln("big fucking eror", err)
-			return false
-		}
-	} else {
-		pow.UpdateCache(blockNum, false)
-		pow.cacheMutex.RLock()
-		defer pow.cacheMutex.RUnlock()
-		pAc = pow.paramsAndCache
-	}
-
-	ret := new(C.ethash_return_value)
-
-	C.ethash_light(ret, pAc.cache, pAc.params, chash, cnonce)
+func (pow *Full) Turbo(on bool) {
+	pow.turbo = on
+}
 
-	result := common.Bytes2Big(C.GoBytes(unsafe.Pointer(&ret.result), C.int(32)))
-	return result.Cmp(target) <= 0
+type Ethash struct {
+	Light
+	Full
 }
 
-func (pow *Ethash) GetHashrate() int64 {
-	return pow.HashRate
+func New() *Ethash {
+	return &Ethash{Light: TheLight}
 }
 
-func (pow *Ethash) Turbo(on bool) {
-	pow.turbo = on
+func NewForTesting() *Ethash {
+	return &Ethash{
+		Light{test: true},
+		Full{test: true},
+	}
 }
 
-func (pow *Ethash) FullHash(nonce uint64, miningHash []byte) []byte {
-	pow.UpdateDAG()
-	pow.dagMutex.Lock()
-	defer pow.dagMutex.Unlock()
-	cMiningHash := (*C.ethash_h256_t)(unsafe.Pointer(&miningHash[0]))
-	cnonce := C.uint64_t(nonce)
-	ret := new(C.ethash_return_value)
-	// pow.hash is the output/return of ethash_full
-	C.ethash_full(ret, pow.dag.dag, pow.paramsAndCache.params, cMiningHash, cnonce)
-	ghash_full := C.GoBytes(unsafe.Pointer(&ret.result), 32)
-	return ghash_full
+func GetSeedHash(blockNum uint64) ([]byte, error) {
+	if blockNum >= epochLength*2048 {
+		return nil, fmt.Errorf("block number is out of bounds (value %v, limit is %v)", blockNum, epochLength*2048)
+	}
+
+	epoch := blockNum / epochLength
+	seedHash := make([]byte, 32)
+	var i uint64
+	for i = 0; i < epoch; i++ {
+		seedHash = crypto.Sha3(seedHash)
+	}
+	return seedHash, nil
 }
 
-func (pow *Ethash) LightHash(nonce uint64, miningHash []byte) []byte {
-	cMiningHash := (*C.ethash_h256_t)(unsafe.Pointer(&miningHash[0]))
-	cnonce := C.uint64_t(nonce)
-	ret := new(C.ethash_return_value)
-	C.ethash_light(ret, pow.paramsAndCache.cache, pow.paramsAndCache.params, cMiningHash, cnonce)
-	ghash_light := C.GoBytes(unsafe.Pointer(&ret.result), 32)
-	return ghash_light
+func (pow *Ethash) Stop() {
+	pow.Light.mu.Lock()
+	pow.Full.mu.Lock()
+	defer pow.Full.mu.Unlock()
```

Comment:
defer in same order as the locks. It's good practice to unlock in reverse order (I think this was your intention, defers work a bit different than you'd expect). The order now is:

```
A.lock
B.lock
A.unlock
b.unlock
```

while I think you're trying to do

```
A.lock
B.lock
B.unlock
A.unlock
```